// ng2-datepicker/dist/pipes/dayClasses.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Pipe } from '@angular/core';
import { DateState } from '../models';
export var DayClassesPipe = (function () {
    function DayClassesPipe() {
    }
    DayClassesPipe.prototype.transform = function (day) {
        var classes = {
            enabled: false,
            inRange: false,
            selected: false,
            today: day.isToday,
            currDisplayMonth: day.isCurrDisplayMonth
        };
        if (day.state > DateState.disabled)
            classes[DateState[day.state]] = true;
        return classes;
    };
    DayClassesPipe = __decorate([
        Pipe({
            name: 'dayClasses'
        }), 
        __metadata('design:paramtypes', [])
    ], DayClassesPipe);
    return DayClassesPipe;
}());
